### PR TITLE
add PDLL mul/div/mod/sub/exp2/log2 operators 

### PR DESCRIPTION
**Pull request by ge28boj**
Features:
1. Allow usage in rewrite section as well. But expression without assignment
`attr<"4 : i32"> + attr<"5 : i32>` is currently forbidden in rewrite section.
2. Support operation based on operator precedence in C/C++:
`let a : Attr = attr<"4 : i32:> + attr<"5 : i32"> * attr<"6 : i32">`
3. No promotion rule so far

### CODE REVIEW

File: mlir/lib/Dialect/PDL/IR/Builtins.cpp [hunk 1]
```diff
@@ -51,8 +51,9 @@ mlir::Attribute addElemToArrayAttr(mlir::PatternRewriter &rewriter,
   return rewriter.getArrayAttr(values);
 }
 
-LogicalResult add(mlir::PatternRewriter &rewriter, mlir::PDLResultList &results,
-                  llvm::ArrayRef<mlir::PDLValue> args) {
+LogicalResult addOrSub(mlir::PatternRewriter &rewriter,
+                       mlir::PDLResultList &results,
+                       llvm::ArrayRef<mlir::PDLValue> args, const bool useAdd) {
```
Comment (ehsan-toosi) [inline mlir/lib/Dialect/PDL/IR/Builtins.cpp:56]: Can you make the `useAdd` a template parameter and use if `constexpr` for checking if it's add/sub?

File: mlir/lib/Dialect/PDL/IR/Builtins.cpp [hunk 2]
```diff
@@ -94,12 +104,160 @@ LogicalResult add(mlir::PatternRewriter &rewriter, mlir::PDLResultList &results,
...
+      return failure();
+    }
+
+    results.push_back(rewriter.getFloatAttr(floatType, resultVal));
+    return success();
+  }
+  return failure();
+}
+
+LogicalResult mulOrDiv(mlir::PatternRewriter &rewriter,
```
Comment (ehsan-toosi) [inline mlir/lib/Dialect/PDL/IR/Builtins.cpp:126]: same

File: mlir/lib/Dialect/PDL/IR/Builtins.cpp [hunk 3]
```diff
@@ -94,12 +104,160 @@ LogicalResult add(mlir::PatternRewriter &rewriter, mlir::PDLResultList &results,
...
+                       mlir::PDLResultList &results,
+                       llvm::ArrayRef<PDLValue> args, const bool useDiv) {
+  assert(args.size() == 2 && "Expected 2 arguments");
```
Comment (ehsan-toosi) [inline mlir/lib/Dialect/PDL/IR/Builtins.cpp:129]: Can the message improve a bit?

File: mlir/lib/Dialect/PDL/IR/Builtins.cpp [hunk 4]
```diff
@@ -94,12 +104,160 @@ LogicalResult add(mlir::PatternRewriter &rewriter, mlir::PDLResultList &results,
...
+      if (useDiv) {
+        resultAPInt = lhsIntAttr.getValue().udiv(rhsIntAttr.getValue());
+      } else {
+        resultAPInt =
+            lhsIntAttr.getValue().umul_ov(rhsIntAttr.getValue(), isOverflow);
+      }
+    } else {
+      if (useDiv) {
+        resultAPInt =
+            lhsIntAttr.getValue().sdiv_ov(rhsIntAttr.getValue(), isOverflow);
```
Comment (ehsan-toosi) [inline mlir/lib/Dialect/PDL/IR/Builtins.cpp:153]: Is division by zero handled by `sdiv_ov`?

File: mlir/lib/Dialect/PDL/IR/Builtins.cpp [hunk 5]
```diff
@@ -94,12 +104,160 @@ LogicalResult add(mlir::PatternRewriter &rewriter, mlir::PDLResultList &results,
...
+      return failure();
+
+    auto integerType = lhsIntAttr.getType();
+
+    llvm::APInt resultAPInt;
+    if (integerType.isSignlessInteger() || integerType.isUnsignedInteger()) {
+      resultAPInt = lhsIntAttr.getValue().urem(rhsIntAttr.getValue());
+    } else {
+      resultAPInt = lhsIntAttr.getValue().srem(rhsIntAttr.getValue());
+    }
```
Comment (ehsan-toosi) [inline mlir/lib/Dialect/PDL/IR/Builtins.cpp:237]: same question for rhs=0. Can we have a test for those as well?

File: mlir/lib/Dialect/PDL/IR/Builtins.cpp [hunk 6]
```diff
@@ -118,6 +276,10 @@ void registerBuiltins(PDLPatternModule &pdlPattern) {
                                      createArrayAttr);
   pdlPattern.registerRewriteFunction("__builtin_addElemToArrayAttr",
                                      addElemToArrayAttr);
+  pdlPattern.registerConstraintFunctionWithResults("__builtin_mul", mul);
+  pdlPattern.registerConstraintFunctionWithResults("__builtin_div", div);
+  pdlPattern.registerConstraintFunctionWithResults("__builtin_mod", mod);
   pdlPattern.registerConstraintFunctionWithResults("__builtin_add", add);
+  pdlPattern.registerConstraintFunctionWithResults("__builtin_sub", sub);
```
Comment (ehsan-toosi) [inline mlir/lib/Dialect/PDL/IR/Builtins.cpp:283]: register as rewrite?

File: mlir/unittests/Dialect/PDL/BuiltinTest.cpp [hunk 7]
```diff
@@ -182,4 +398,101 @@ TEST_F(BuiltinTest, add) {
     EXPECT_TRUE(builtin::add(rewriter, results, {zerof32, zerof64}).failed());
   }
 }
+
+TEST_F(BuiltinTest, sub) {
```
Comment (ehsan-toosi) [inline mlir/unittests/Dialect/PDL/BuiltinTest.cpp:402]: Will it make sense to properly test every data type (or at least the types we are using now) like [s/u]i8, [s/u]i32, [s/u]i64, f16, f32, f64 for corner cases (overflow, underflow, division by zero, ...) and correctness? or you believe because you are using [u/s]mul/rem/udiv/ from APINT/APfloat, it's not necessary.
In anycase, please add a test for division by zero.

**Review by ehsan-toosi — APPROVED** [review]

File: mlir/lib/Tools/PDLL/Parser/Lexer.cpp [hunk 8]
```diff
@@ -139,7 +139,7 @@ int Lexer::getNextChar() {
     return static_cast<unsigned char>(curChar);
   case 0: {
     // A nul character in the stream is either the end of the current buffer
-    // or a random nul in the file. Disambiguate that here.
+    // or a random nul in the file. Disambiguate that here.r
```
Comment (ttjost) [inline mlir/lib/Tools/PDLL/Parser/Lexer.cpp:142]:
```suggestion
    // or a random nul in the file. Disambiguate that here.
```

File: mlir/lib/Tools/PDLL/Parser/Parser.cpp [hunk 9]
```diff
@@ -2559,9 +2760,12 @@ FailureOr<ast::Stmt *> Parser::parseStmt(bool expectTerminalSemicolon) {
   case Token::kw_erase:
     stmt = parseEraseStmt();
     break;
-  case Token::kw_let:
+  case Token::kw_let: {
     stmt = parseLetStmt();
+    llvm::SaveAndRestore saveCtx(nativeOperatorContext,
```
Comment (ttjost) [inline mlir/lib/Tools/PDLL/Parser/Parser.cpp:2765]: Isn't `SaveAndRestore` doing RAII? I wonder if we still need to set it back to `Generic`, since the value should be restored once it leaves `parseLetStmt`.

File: mlir/lib/Tools/PDLL/Parser/Parser.cpp [hunk 10]
```diff
@@ -1915,28 +1964,180 @@ FailureOr<ast::Expr *> Parser::parseEqualityExpr() {
...
+    consumeToken();
+    consumeToken(Token::l_paren);
+    expr = parseAddSubExpr();
+    if (failed(expr))
+      return failure();
+
+    // Check if it is in rewrite section but not in the let statement
+    bool inRewriteSection = parserContext == ParserContext::Rewrite;
+    if (inRewriteSection &&
+        nativeOperatorContext == NativeOperatorContext::Generic)
```
Comment (ttjost) [inline mlir/lib/Tools/PDLL/Parser/Parser.cpp:2104]: Can't we check that this isn't a `Let`?
```suggestion
        nativeOperatorContext != NativeOperatorContext::Let)
```

File: mlir/lib/Tools/PDLL/Parser/Parser.cpp [hunk 11]
```diff
@@ -1915,28 +1964,180 @@ FailureOr<ast::Expr *> Parser::parseEqualityExpr() {
...
+      consumeToken();
+      auto rhs = parseMulDivModExpr();
+      if (failed(rhs))
+        return failure();
+      SmallVector<ast::Expr *> args{*lhs, *rhs};
+
+      // Check if it is in rewrite section but not in the let statement
+      bool inRewriteSection = parserContext == ParserContext::Rewrite;
+      if (inRewriteSection &&
+          nativeOperatorContext == NativeOperatorContext::Generic)
```
Comment (ttjost) [inline mlir/lib/Tools/PDLL/Parser/Parser.cpp:1983]: Can't we check that this isn't a Let?
```suggestion
          nativeOperatorContext == NativeOperatorContext::Generic)
```

File: mlir/lib/Tools/PDLL/Parser/Parser.cpp [hunk 12]
```diff
@@ -1915,28 +1964,180 @@ FailureOr<ast::Expr *> Parser::parseEqualityExpr() {
...
+        return emitError(
+            "nodiscard rule for add operator is applied in rewrite section");
```
Comment (ttjost) [inline mlir/lib/Tools/PDLL/Parser/Parser.cpp:1985]: I think this message could be improved. Maybe something like:
```suggestion
            "cannot evaluate add operator in rewrite section. Assign to a variable with `let`");
```
I am not very fan with my suggestion, so feel free to improve it. 😄  maybe someone else has a better suggestion.

**Review by ttjost — APPROVED** [review]
Overall, LGTM.

File: mlir/lib/Dialect/PDL/IR/Builtins.cpp [hunk 13]
```diff
@@ -51,52 +52,172 @@ mlir::Attribute addElemToArrayAttr(mlir::PatternRewriter &rewriter,
...
+      return failure();
+    }
+    return success();
+  }
+
+  if (auto operandFloatAttr = dyn_cast_or_null<FloatAttr>(operandAttr)) {
+    auto floatType = operandFloatAttr.getType();
+
+    if constexpr (T == UnaryOpKind::exp2) {
+      auto maxVal = APFloat::getLargest(llvm::APFloat::IEEEhalf());
```
Comment (mgehre-amd) [inline mlir/lib/Dialect/PDL/IR/Builtins.cpp:108]: Why do we use `IEEEhalf` here?

File: mlir/lib/Tools/PDLL/Parser/Parser.cpp [hunk 14]
```diff
@@ -1915,28 +1964,180 @@ FailureOr<ast::Expr *> Parser::parseEqualityExpr() {
 FailureOr<ast::Expr *> Parser::parseRelationExpr() { return parseAddSubExpr(); }
 
 FailureOr<ast::Expr *> Parser::parseAddSubExpr() {
-  auto lhs = parseMulDivExpr();
+  auto lhs = parseMulDivModExpr();
+  if (failed(lhs))
+    return failure();
+
+  for (;;) {
```
Comment (mgehre-amd) [inline mlir/lib/Tools/PDLL/Parser/Parser.cpp:1971]: Why does this need a loop?

Comment (ge28boj) [reply mlir/lib/Tools/PDLL/Parser/Parser.cpp:1971]: It enables multiple operations of same precedence, e.g. `let a : Attr = attr<"4 : i32"> + attr<"5 : i32"> - attr<"6 : i32">`. Does it make sense to you?

File: mlir/lib/Dialect/PDL/IR/Builtins.cpp (hunk 13 above)
Comment (ge28boj) [reply mlir/lib/Dialect/PDL/IR/Builtins.cpp:108]: Thanks for catching it. It should be handled case by case depending on float type.

File: mlir/lib/Dialect/PDL/IR/Builtins.cpp [hunk 15]
```diff
@@ -51,52 +52,172 @@ mlir::Attribute addElemToArrayAttr(mlir::PatternRewriter &rewriter,
...
+    } else if constexpr (T == BinaryOpKind::mod) {
+      if (integerType.isSignlessInteger() || integerType.isUnsignedInteger()) {
+        resultAPInt = lhsIntAttr.getValue().urem(rhsIntAttr.getValue());
+      } else {
+        resultAPInt = lhsIntAttr.getValue().srem(rhsIntAttr.getValue());
+      }
     } else {
-      resultAPInt =
-          lhsIntAttr.getValue().sadd_ov(rhsIntAttr.getValue(), isOverflow);
+      assert(false && "Unsupported binary operator");
```
Comment (ehsan-toosi) [inline mlir/lib/Dialect/PDL/IR/Builtins.cpp:180]: Why assert and not an llvm erro?

File: mlir/lib/Dialect/PDL/IR/Builtins.cpp [hunk 16]
```diff
@@ -51,52 +52,172 @@ mlir::Attribute addElemToArrayAttr(mlir::PatternRewriter &rewriter,
...
+  auto operandAttr = args[0].cast<Attribute>();
+
+  if (auto operandIntAttr = dyn_cast_or_null<IntegerAttr>(operandAttr)) {
+    auto integerType = cast<IntegerType>(operandIntAttr.getType());
+    auto bitWidth = integerType.getIntOrFloatBitWidth();
+
+    if constexpr (T == UnaryOpKind::exp2) {
+      uint64_t resultVal =
+          integerType.isUnsigned() || integerType.isSignless()
+              ? std::pow(2, operandIntAttr.getValue().getZExtValue())
```
Comment (ehsan-toosi) [inline mlir/lib/Dialect/PDL/IR/Builtins.cpp:68]: Is this operator what we need actually? Not for this PR but wouldn't it better to have pow instead of this? or even `exp`?

**Review by ehsan-toosi — APPROVED** [review]
LG. Thanks.

File: mlir/lib/Tools/PDLL/Parser/Parser.cpp (hunk 14 above)
Comment (mgehre-amd) [reply mlir/lib/Tools/PDLL/Parser/Parser.cpp:1971]: I think you can achieve the same by doing `rhs = parseAddSubExpr()` below,
which, from what I understand, is the typical way how recursive descent parsers are implemented.
Would that work?
